### PR TITLE
retroarch: refactor

### PR DESCRIPTION
**Pull request by thiagokokada**
- Move all RetroArch related packages to `pkgs/by-name`
- Rename `retroarchFull` to `retroarch-full` and `retroarchBare` to `retroarch-bare` (also add aliases to the old names)
- Add `retroarch.withCores` helper to make it easier to add cores to RetroArch and it is more inline with e.g.: `python3.withPackages`
- Add `retroarch-free`, a derivation of RetroArch with only free cores (e.g.: can be built in OfBorg)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by thiagokokada** [comment]
@ofborg build retroarch-free

File: pkgs/by-name/re/retroarch-bare/package.nix [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  meta = with lib; {
+    homepage = "https://libretro.com";
+    description = "Multi-platform emulator frontend for libretro cores";
+    license = licenses.gpl3Plus;
+    platforms = platforms.unix;
+    changelog = "https://github.com/libretro/RetroArch/blob/v${version}/CHANGES.md";
+    maintainers =
+      with maintainers;
+      teams.libretro.members
```
Comment (AndersonTorres) [inline pkgs/by-name/re/retroarch-bare/package.nix:196]:
```suggestion
      teams.libretro.members
      ++ with maintainers;
```

File: pkgs/by-name/re/retroarch-bare/package.nix [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+          symlinkJoin
+          cores
+          ;
+        settings = {
+          joypad_autoconfig_dir = "${retroarch-joypad-autoconfig}/share/libretro/autoconfig";
+        } // settings;
+      };
+  };
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/by-name/re/retroarch-bare/package.nix:188]:
```suggestion
  meta = {
```

File: pkgs/by-name/re/retroarch-bare/package.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/by-name/re/retroarch-bare/package.nix:196]: Doesn't work, it is a compile error.

Comment (thiagokokada) [reply pkgs/by-name/re/retroarch-bare/package.nix]: Reversed the order so it compiles.

**Comment by thiagokokada** [comment]
@ofborg build retroarch-free

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358405`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 103 packages built:</summary>
  <ul>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>libretro.atari800</li>
    <li>libretro.beetle-gba</li>
    <li>libretro.beetle-lynx</li>
    <li>libretro.beetle-ngp</li>
    <li>libretro.beetle-pce</li>
    <li>libretro.beetle-pce-fast</li>
    <li>libretro.beetle-pcfx</li>
    <li>libretro.beetle-psx</li>
    <li>libretro.beetle-psx-hw</li>
    <li>libretro.beetle-saturn</li>
    <li>libretro.beetle-supafaust</li>
    <li>libretro.beetle-supergrafx</li>
    <li>libretro.beetle-vb</li>
    <li>libretro.beetle-wswan</li>
    <li>libretro.blastem</li>
    <li>libretro.bluemsx</li>
    <li>libretro.bsnes</li>
    <li>libretro.bsnes-hd</li>
    <li>libretro.bsnes-mercury</li>
    <li>libretro.bsnes-mercury-balanced</li>
    <li>libretro.bsnes-mercury-performance</li>
    <li>libretro.citra</li>
    <li>libretro.desmume</li>
    <li>libretro.desmume2015</li>
    <li>libretro.dolphin</li>
    <li>libretro.dosbox</li>
    <li>libretro.dosbox-pure</li>
    <li>libretro.easyrpg</li>
    <li>libretro.eightyone</li>
    <li>libretro.fbalpha2012</li>
    <li>libretro.fbneo</li>
    <li>libretro.fceumm</li>
    <li>libretro.flycast</li>
    <li>libretro.fmsx</li>
    <li>libretro.freeintv</li>
    <li>libretro.fuse</li>
    <li>libretro.gambatte</li>
    <li>libretro.genesis-plus-gx</li>
    <li>libretro.gpsp</li>
    <li>libretro.gw</li>
    <li>libretro.handy</li>
    <li>libretro.hatari</li>
    <li>libretro.mame</li>
    <li>libretro.mame2000</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2003-plus</li>
    <li>libretro.mame2010</li>
    <li>libretro.mame2015</li>
    <li>libretro.mame2016</li>
    <li>libretro.melonds</li>
    <li>libretro.mesen</li>
    <li>libretro.mesen-s</li>
    <li>libretro.meteor</li>
    <li>libretro.mgba</li>
    <li>libretro.mrboom</li>
    <li>libretro.mupen64plus</li>
    <li>libretro.neocd</li>
    <li>libretro.nestopia</li>
    <li>libretro.np2kai</li>
    <li>libretro.nxengine</li>
    <li>libretro.o2em</li>
    <li>libretro.opera</li>
    <li>libretro.parallel-n64</li>
    <li>libretro.pcsx-rearmed (libretro.pcsx_rearmed)</li>
    <li>libretro.pcsx2</li>
    <li>libretro.picodrive</li>
    <li>libretro.play</li>
    <li>libretro.ppsspp</li>
    <li>libretro.prboom</li>
    <li>libretro.prosystem</li>
    <li>libretro.puae</li>
    <li>libretro.quicknes</li>
    <li>libretro.same_cdi</li>
    <li>libretro.sameboy</li>
    <li>libretro.scummvm</li>
    <li>libretro.smsplus-gx</li>
    <li>libretro.snes9x</li>
    <li>libretro.snes9x2002</li>
    <li>libretro.snes9x2005</li>
    <li>libretro.snes9x2005-plus</li>
    <li>libretro.snes9x2010</li>
    <li>libretro.stella</li>
    <li>libretro.stella2014</li>
    <li>libretro.swanstation</li>
    <li>libretro.tgbdual</li>
    <li>libretro.thepowdertoy</li>
    <li>libretro.tic80</li>
    <li>libretro.twenty-fortyeight</li>
    <li>libretro.vba-m</li>
    <li>libretro.vba-next</li>
    <li>libretro.vecx</li>
    <li>libretro.virtualjaguar</li>
    <li>libretro.yabause</li>
    <li>retroarch</li>
    <li>retroarch-bare</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
CC @aanderse for review.

**Review by gepbird — DISMISSED** [review]
[`fa080ef` (#358405)](https://github.com/NixOS/nixpkgs/pull/358405/commits/fa080efee76ccb7b54fe80b3c1b57e0b4ed2e74a) contains more changes than just a by-name move, maybe you should split up the package rename changes.
Similar things with [`31dc6a6` (#358405)](https://github.com/NixOS/nixpkgs/pull/358405/commits/31dc6a6dcb3e025098b5f2a31387e2ecc91e6706).

**Comment by thiagokokada** [comment]
> [`fa080ef` (#358405)](https://github.com/NixOS/nixpkgs/pull/358405/commits/fa080efee76ccb7b54fe80b3c1b57e0b4ed2e74a) contains more changes than just a by-name move, maybe you should split up the package rename changes. Similar things with [`31dc6a6` (#358405)](https://github.com/NixOS/nixpkgs/pull/358405/commits/31dc6a6dcb3e025098b5f2a31387e2ecc91e6706).

Done.

**Comment by thiagokokada** [comment]
@ofborg build retroarch-free

File: pkgs/applications/emulators/libretro/README.md [hunk 3]
```diff
@@ -15,24 +15,24 @@ The basic steps to add a new core are:
 ## Using RetroArch with cores
 
 To create a custom RetroArch derivation with the cores you want (instead of
-using `retroarchFull` that includes all cores), you can use `.override` like
+using `retroarch-full` that includes all cores), you can use `.override` like
```
Comment (aanderse) [inline pkgs/applications/emulators/libretro/README.md:18]:
```suggestion
using `retroarch-full` that includes all cores), you can use `.withCores` like
```

**Review by aanderse — APPROVED** [review]
very awesome!

**Review by gepbird — APPROVED** [review]
Thanks, LGTM!

File: pkgs/applications/emulators/libretro/README.md (hunk 3 above)
Comment (thiagokokada) [reply pkgs/applications/emulators/libretro/README.md]: Done.

**Comment by thiagokokada** [comment]
@ofborg build retroarch-free

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358405`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 103 packages built:</summary>
  <ul>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>libretro.atari800</li>
    <li>libretro.beetle-gba</li>
    <li>libretro.beetle-lynx</li>
    <li>libretro.beetle-ngp</li>
    <li>libretro.beetle-pce</li>
    <li>libretro.beetle-pce-fast</li>
    <li>libretro.beetle-pcfx</li>
    <li>libretro.beetle-psx</li>
    <li>libretro.beetle-psx-hw</li>
    <li>libretro.beetle-saturn</li>
    <li>libretro.beetle-supafaust</li>
    <li>libretro.beetle-supergrafx</li>
    <li>libretro.beetle-vb</li>
    <li>libretro.beetle-wswan</li>
    <li>libretro.blastem</li>
    <li>libretro.bluemsx</li>
    <li>libretro.bsnes</li>
    <li>libretro.bsnes-hd</li>
    <li>libretro.bsnes-mercury</li>
    <li>libretro.bsnes-mercury-balanced</li>
    <li>libretro.bsnes-mercury-performance</li>
    <li>libretro.citra</li>
    <li>libretro.desmume</li>
    <li>libretro.desmume2015</li>
    <li>libretro.dolphin</li>
    <li>libretro.dosbox</li>
    <li>libretro.dosbox-pure</li>
    <li>libretro.easyrpg</li>
    <li>libretro.eightyone</li>
    <li>libretro.fbalpha2012</li>
    <li>libretro.fbneo</li>
    <li>libretro.fceumm</li>
    <li>libretro.flycast</li>
    <li>libretro.fmsx</li>
    <li>libretro.freeintv</li>
    <li>libretro.fuse</li>
    <li>libretro.gambatte</li>
    <li>libretro.genesis-plus-gx</li>
    <li>libretro.gpsp</li>
    <li>libretro.gw</li>
    <li>libretro.handy</li>
    <li>libretro.hatari</li>
    <li>libretro.mame</li>
    <li>libretro.mame2000</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2003-plus</li>
    <li>libretro.mame2010</li>
    <li>libretro.mame2015</li>
    <li>libretro.mame2016</li>
    <li>libretro.melonds</li>
    <li>libretro.mesen</li>
    <li>libretro.mesen-s</li>
    <li>libretro.meteor</li>
    <li>libretro.mgba</li>
    <li>libretro.mrboom</li>
    <li>libretro.mupen64plus</li>
    <li>libretro.neocd</li>
    <li>libretro.nestopia</li>
    <li>libretro.np2kai</li>
    <li>libretro.nxengine</li>
    <li>libretro.o2em</li>
    <li>libretro.opera</li>
    <li>libretro.parallel-n64</li>
    <li>libretro.pcsx-rearmed (libretro.pcsx_rearmed)</li>
    <li>libretro.pcsx2</li>
    <li>libretro.picodrive</li>
    <li>libretro.play</li>
    <li>libretro.ppsspp</li>
    <li>libretro.prboom</li>
    <li>libretro.prosystem</li>
    <li>libretro.puae</li>
    <li>libretro.quicknes</li>
    <li>libretro.same_cdi</li>
    <li>libretro.sameboy</li>
    <li>libretro.scummvm</li>
    <li>libretro.smsplus-gx</li>
    <li>libretro.snes9x</li>
    <li>libretro.snes9x2002</li>
    <li>libretro.snes9x2005</li>
    <li>libretro.snes9x2005-plus</li>
    <li>libretro.snes9x2010</li>
    <li>libretro.stella</li>
    <li>libretro.stella2014</li>
    <li>libretro.swanstation</li>
    <li>libretro.tgbdual</li>
    <li>libretro.thepowdertoy</li>
    <li>libretro.tic80</li>
    <li>libretro.twenty-fortyeight</li>
    <li>libretro.vba-m</li>
    <li>libretro.vba-next</li>
    <li>libretro.vecx</li>
    <li>libretro.virtualjaguar</li>
    <li>libretro.yabause</li>
    <li>retroarch</li>
    <li>retroarch-bare</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Comment by bjornfor** [comment]
> The reason for this is because when you open a content and close it
(using `F1` -> `Close content`), RetroArch will reload the configs but
will ignore `--appendconfig`. This definitely looks like a bug, but
hardcoding the values works.

Submit issue upstream and link back here?

**Comment by thiagokokada** [comment]
> > The reason for this is because when you open a content and close it
> > (using `F1` -> `Close content`), RetroArch will reload the configs but
> > will ignore `--appendconfig`. This definitely looks like a bug, but
> > hardcoding the values works.
> 
> Submit issue upstream and link back here?

It could be an issue with our wrapper or something else we do though, this is why I prefer to not report it right now.

If I can reproduce this with vanilla RetroArch I could report it, but unless someone is willing to do the report I don't have time right now.

**Comment by thiagokokada** [comment]
> > The reason for this is because when you open a content and close it
> > (using `F1` -> `Close content`), RetroArch will reload the configs but
> > will ignore `--appendconfig`. This definitely looks like a bug, but
> > hardcoding the values works.
> 
> Submit issue upstream and link back here?

Actually testing this it doesn't solve the issue. Let me drop this commit because otherwise this complicates the derivation for no good reason.

**Comment by thiagokokada** [comment]
@ofborg build retroarch-free

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 358405`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 103 packages built:</summary>
  <ul>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>libretro.atari800</li>
    <li>libretro.beetle-gba</li>
    <li>libretro.beetle-lynx</li>
    <li>libretro.beetle-ngp</li>
    <li>libretro.beetle-pce</li>
    <li>libretro.beetle-pce-fast</li>
    <li>libretro.beetle-pcfx</li>
    <li>libretro.beetle-psx</li>
    <li>libretro.beetle-psx-hw</li>
    <li>libretro.beetle-saturn</li>
    <li>libretro.beetle-supafaust</li>
    <li>libretro.beetle-supergrafx</li>
    <li>libretro.beetle-vb</li>
    <li>libretro.beetle-wswan</li>
    <li>libretro.blastem</li>
    <li>libretro.bluemsx</li>
    <li>libretro.bsnes</li>
    <li>libretro.bsnes-hd</li>
    <li>libretro.bsnes-mercury</li>
    <li>libretro.bsnes-mercury-balanced</li>
    <li>libretro.bsnes-mercury-performance</li>
    <li>libretro.citra</li>
    <li>libretro.desmume</li>
    <li>libretro.desmume2015</li>
    <li>libretro.dolphin</li>
    <li>libretro.dosbox</li>
    <li>libretro.dosbox-pure</li>
    <li>libretro.easyrpg</li>
    <li>libretro.eightyone</li>
    <li>libretro.fbalpha2012</li>
    <li>libretro.fbneo</li>
    <li>libretro.fceumm</li>
    <li>libretro.flycast</li>
    <li>libretro.fmsx</li>
    <li>libretro.freeintv</li>
    <li>libretro.fuse</li>
    <li>libretro.gambatte</li>
    <li>libretro.genesis-plus-gx</li>
    <li>libretro.gpsp</li>
    <li>libretro.gw</li>
    <li>libretro.handy</li>
    <li>libretro.hatari</li>
    <li>libretro.mame</li>
    <li>libretro.mame2000</li>
    <li>libretro.mame2003</li>
    <li>libretro.mame2003-plus</li>
    <li>libretro.mame2010</li>
    <li>libretro.mame2015</li>
    <li>libretro.mame2016</li>
    <li>libretro.melonds</li>
    <li>libretro.mesen</li>
    <li>libretro.mesen-s</li>
    <li>libretro.meteor</li>
    <li>libretro.mgba</li>
    <li>libretro.mrboom</li>
    <li>libretro.mupen64plus</li>
    <li>libretro.neocd</li>
    <li>libretro.nestopia</li>
    <li>libretro.np2kai</li>
    <li>libretro.nxengine</li>
    <li>libretro.o2em</li>
    <li>libretro.opera</li>
    <li>libretro.parallel-n64</li>
    <li>libretro.pcsx-rearmed (libretro.pcsx_rearmed)</li>
    <li>libretro.pcsx2</li>
    <li>libretro.picodrive</li>
    <li>libretro.play</li>
    <li>libretro.ppsspp</li>
    <li>libretro.prboom</li>
    <li>libretro.prosystem</li>
    <li>libretro.puae</li>
    <li>libretro.quicknes</li>
    <li>libretro.same_cdi</li>
    <li>libretro.sameboy</li>
    <li>libretro.scummvm</li>
    <li>libretro.smsplus-gx</li>
    <li>libretro.snes9x</li>
    <li>libretro.snes9x2002</li>
    <li>libretro.snes9x2005</li>
    <li>libretro.snes9x2005-plus</li>
    <li>libretro.snes9x2010</li>
    <li>libretro.stella</li>
    <li>libretro.stella2014</li>
    <li>libretro.swanstation</li>
    <li>libretro.tgbdual</li>
    <li>libretro.thepowdertoy</li>
    <li>libretro.tic80</li>
    <li>libretro.twenty-fortyeight</li>
    <li>libretro.vba-m</li>
    <li>libretro.vba-next</li>
    <li>libretro.vecx</li>
    <li>libretro.virtualjaguar</li>
    <li>libretro.yabause</li>
    <li>retroarch</li>
    <li>retroarch-bare</li>
    <li>retroarch-free</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Pull request merged**